src/backend/distributed/worker/worker_shard_visibility.c

```diff
@@ -0,0 +1,299 @@
+ * any distributed table in the current search path.
+ *
+ * We can only do that in MX since both the metadata and tables are only
+ * present there.
+ */
+static bool
+RelationIsAKnownShard(Oid shardRelationId)
+{
+	int localGroupId = -1;
+	char *shardRelationName = NULL;
+	char *relationName = NULL;
+	bool missingOk = true;
+	uint64 shardId = INVALID_SHARD_ID;
+	Oid relationId = InvalidOid;
+	char *shardIdString = NULL;
+	int relationNameLength = 0;
+	char relKind = '\0';
+
+	List *shardIdList = NIL;
+	ListCell *shardIdCell = NULL;
+
+	if (!OidIsValid(shardRelationId))
+	{
+		/* we cannot continue without a valid Oid */
+		return false;
+	}
+
+	localGroupId = GetLocalGroupId();
+	if (localGroupId == 0)
+	{
+		/*
+		 * We're not interested in shards in the coordinator
+		 * or non-mx worker nodes.
+		 */
+		return false;
+	}
+
+	/* we're not interested in the relations that are not in the search path */
+	if (!RelationIsVisible(shardRelationId))
+	{
+		return false;
+	}
+
+	/*
+	 * If the input relation is an index we simply replace the
+	 * relationId with the corresponding relation to hide indexes
+	 * as well.
+	 */
+	relKind = get_rel_relkind(shardRelationId);
+	if (relKind == RELKIND_INDEX)
+	{
+		shardRelationId = IndexGetRelation(shardRelationId, false);
+	}
+
+	/* get the shard's relation name */
+	shardRelationName = get_rel_name(shardRelationId);
+
+	/* find the last underscore for shardId string */
+	shardIdString = strrchr(shardRelationName, SHARD_NAME_SEPARATOR);
+	if (shardIdString == NULL)
+	{
+		/* there are no underscore in the table name */
+		return false;
+	}
+
+	relationNameLength = shardIdString - shardRelationName;
+	relationName = citus_strndup(shardRelationName, relationNameLength);
+
+	relationId = RelnameGetRelid(relationName);
+	if (!OidIsValid(relationId))
+	{
+		/* there is no such relation */
+		return false;
+	}
+
+	if (!IsDistributedTable(relationId))
+	{
+		/* we're  obviously only interested in distributed tables */
+		return false;
+	}
+
+	shardId = ExtractShardId(shardRelationName, missingOk);
+	if (shardId == INVALID_SHARD_ID)
+	{
+		/*
+		 * The format of the table name does not align with
+		 * our shard name definition.
+		 */
+		return false;
+	}
+
+	/*
+	 * Finally make sure that the input shard belongs to
+	 * the distributed table we've found.
+	 */
+	shardIdList = LoadShardList(relationId);
+	foreach(shardIdCell, shardIdList)
+	{
+		uint64 *shardIdPointer = (uint64 *) lfirst(shardIdCell);
+		uint64 currentShardId = (*shardIdPointer);
+
+		if (shardId == currentShardId)
+		{
+			return true;
+		}
+	}
+
+	return false;
+}
+
+
+/*
+ * strndup is not part of Standard C and extension of Posix.1-2008.
+ * So, implement our own strndup to make it work on Windows.
+ */
+static char *
+citus_strndup(const char *s, size_t n)
+{
+	char *result;
+	size_t len = strlen(s);
+
+	if (n < len)
+	{
+		len = n;
+	}
+
+	result = (char *) malloc(len + 1);
```

Comment:
do we need to care about deallocation and memory context ?

Reply:
Oh, that's a good catch. We should simply use `palloc`, right?

Reply:
Why not `pstrdup`?

Reply:
Interesting, I searched for `pstrndup` in Postgres source, but the function is named  `pnstrdup`. So, I'll switch to that.

(note that I don't want to use `pstrdup ` because the goal here is to get the main table name from the shard name, where we know the exact length we want to copy: `relationName = citus_strndup(shardRelationName, relationNameLength);`)

Reply:
What if the relation name is truncated because it is very long?

Reply:
e.g. https://github.com/citusdata/citus/blob/master/src/test/regress/expected/multi_name_lengths.out#L29

Reply:
Ops, it is currently broken with this PR and that's something slightly hard to solve. We should do something similar [here](https://github.com/citusdata/citus/commit/f677fadbe686ed370844549e06d9cd3c8c46bbb0#diff-41af17ffe3adc232e63ff6f138e3ab74R648). But, here we should iterate on all distributed relations and find the distributed table with the same hash when the `shardRelationName` long.

Reply:
Could we iterate over the shards, construct the name, and include them in the result if there is a table with that name?

Reply:
I decided to not rely on table/shard names at all. The new algorithm is simpler with some overhead: Extract shard id -> Read `pg_dist_shard` (via index scan on shardId)** -> Make sure that the shard belongs to a distributed table.

This covers the long shard names and allows us to get rid of `citus_strndup ` function.

** note that the overhead is neglicable, I tested locally with 4000 shards on a worker.